Code/GraphMol/SmilesParse/CXSmilesOps.cpp

```diff
@@ -250,6 +250,249 @@ bool parse_coordinate_bonds(Iterator &first, Iterator last, RDKit::RWMol &mol) {
   return true;
 }
 
+template <typename Iterator>
+bool parse_unsaturation(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  if (first + 1 >= last || *first != 'u') {
+    return false;
+  }
+  ++first;
+  if (first >= last || *first != ':') {
+    return false;
+  }
+  ++first;
+  while (first < last && *first >= '0' && *first <= '9') {
+    unsigned int idx;
+    if (!read_int(first, last, idx)) {
+      return false;
+    }
+    auto atom = mol.getAtomWithIdx(idx);
+    if (!atom->hasQuery()) {
+      atom = QueryOps::replaceAtomWithQueryAtom(&mol, atom);
+    }
+    atom->expandQuery(makeAtomUnsaturatedQuery(), Queries::COMPOSITE_AND);
+    if (first < last && *first == ',') {
+      ++first;
+    }
+  }
+  return true;
+}
+
+template <typename Iterator>
+bool parse_ring_bonds(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  if (first >= last || *first != 'r' || first + 1 >= last ||
+      *(first + 1) != 'b' || first + 2 >= last || *(first + 2) != ':') {
+    return false;
+  }
+  first += 3;
+  while (first < last && *first >= '0' && *first <= '9') {
+    unsigned int n1;
+    if (!read_int(first, last, n1)) {
+      return false;
+    }
+    // check that we can read at least two more characters:
+    if (first + 1 >= last || *first != ':') {
+      return false;
+    }
+    ++first;
+    unsigned int n2;
+    bool gt = false;
+    if (*first == '*') {
+      ++first;
+      n2 = 0xDEADBEEF;
+      mol.setProp(common_properties::_NeedsQueryScan, 1);
+    } else {
+      if (!read_int(first, last, n2)) {
+        return false;
+      }
+      switch (n2) {
+        case 0:
+        case 2:
+        case 3:
+          break;
+        case 4:
+          gt = true;
+          break;
+        default:
+          BOOST_LOG(rdWarningLog)
+              << "unrecognized rb value: " << n2 << std::endl;
+          return false;
+      }
+    }
+    auto atom = mol.getAtomWithIdx(n1);
+    if (!atom->hasQuery()) {
+      atom = QueryOps::replaceAtomWithQueryAtom(&mol, atom);
+    }
+    if (!gt) {
+      atom->expandQuery(makeAtomRingBondCountQuery(n2), Queries::COMPOSITE_AND);
+    } else {
+      auto q = static_cast<ATOM_EQUALS_QUERY *>(new ATOM_LESSEQUAL_QUERY);
+      q->setVal(n2);
+      q->setDescription("AtomRingBondCount");
+      q->setDataFunc(queryAtomRingBondCount);
+      atom->expandQuery(q, Queries::COMPOSITE_AND);
+    }
+    if (first < last && *first == ',') {
+      ++first;
+    }
+  }
+  return true;
+}
+
+template <typename Iterator>
+bool parse_linknodes(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  // these look like: |LN:1:1.3.2.6,4:1.4.3.6|
+  // that's two records:
+  //   1:1.3.2.6: 1-3 repeats, atom 1-2, 1-6
+  //   4:1.4.3.6: 1-4 repeats, atom 4-4, 4-6
+  // which maps to the property value "1 3 2 2 3 2 7|1 4 2 5 4 5 7"
```

Comment:
I checked the [CXSMILES docs for link nodes](https://docs.chemaxon.com/display/docs/ChemAxon_Extended_SMILES_and_SMARTS_-_CXSMILES_and_CXSMARTS.html), and it says this:

> If the link node has only two connections, then the first and second outer atom indexes are obvious, so they are omitted.

So, apparently, in some cases we can have link nodes with just two "parameters", like `|LN:1:1.3|`. Should we support these too?

Reply:
In principle yes, but when I was writing the code I couldn't figure out what that would actually mean and couldn't make marvin generate an example. Now I think I get it and I will update the code and add some tests.

Reply:
Makes sense.